File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans;
    Pair<Integer, Integer> solve(TreeNode node) {
        if(node == null) return new Pair<>(0, 0);
        var p1 = solve(node.left);
        var p2 = solve(node.right);
        int sum = p1.getKey() + p2.getKey() + node.val;
        int nodes = p1.getValue() + p2.getValue() + 1;
        if(sum / nodes == node.val) ans++;
        return new Pair<>(sum, nodes);
    }
    public int averageOfSubtree(TreeNode root) {
        ans = 0;
        solve(root);
        return ans;
    }
}

File: 2347-count-nodes-equal-to-average-of-subtree/count-nodes-equal-to-average-of-subtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var averageOfSubtree = function(root) {
    let ans = 0;

    function dfs(node) {
        if(node === null) return [0, 0];
        let left = dfs(node.left);
        let right = dfs(node.right);
        let sum = left[0] + right[0] + node.val;
        let nodes = left[1] + right[1] + 1;
        if(Math.floor(sum / nodes) === node.val) ans++;
        return [sum, nodes];
    }

    dfs(root);

    return ans;
};